### PR TITLE
Delete peer details from store after peer leaves cluster

### LINKED ISSUE #1108: Peer Deletion Behavior  if GD2 is not running on Node
Deletion of a Node on which GD2 is not running results in failure. But If I do a get peers request then that node is not present in the list.
I am expecting that node to be present in the list, only online status should be false.

## Steps:
For a 3 node cluster setup and using external etcd
##### Response
&nbsp;

```javascript
[
    {
        "id": "2f0bfdac-fceb-459a-8939-689da17de77e",
        "name": "gd21.example",
        "peer-addresses": [
           "192.168.122.6:24008"
        ],
        "client-addresses": [
            "127.0.0.1:24007",
            "192.168.122.6:24007"
        ],
        "online": true,
        "pid": 5475,
        "metadata": {
            "_zone": "2f0bfdac-fceb-459a-8939-689da17de77e"
        }
    },
    {
        "id": "854e037e-da80-4ee0-82b8-9ecdb8fc13de",
        "name": "gd22.example",
        "peer-addresses": [
            "192.168.122.42:24008"
        ],
        "client-addresses": [
            "127.0.0.1:24007",
            "192.168.122.42:24007"
        ],
        "online": true,
        "pid": 4896,
        "metadata": {
            "_zone": "854e037e-da80-4ee0-82b8-9ecdb8fc13de"
        }
    },
    {
        "id": "a35e9c07-fc7c-45ac-a0a3-4e09869b30ca",
        "name": "gd20.example",
        "peer-addresses": [
            "192.168.122.253:24008"
        ],
        "client-addresses": [
            "127.0.0.1:24007",
            "192.168.122.253:24007"
        ],
        "online": true,
        "pid": 4982,
        "metadata": {
            "_zone": "a35e9c07-fc7c-45ac-a0a3-4e09869b30ca"
        }
    }
]
```
#### 3. Get Peers
```javascript
[
    {
        "id": "2f0bfdac-fceb-459a-8939-689da17de77e",
        "name": "gd21.example",
        "peer-addresses": [
            "192.168.122.6:24008"
        ],
        "client-addresses": [
            "127.0.0.1:24007",
            "192.168.122.6:24007"
        ],
        "online": false,
        "metadata": {
            "_zone": "2f0bfdac-fceb-459a-8939-689da17de77e"
        }
    },
    {
        "id": "854e037e-da80-4ee0-82b8-9ecdb8fc13de",
        "name": "gd22.example",
        "peer-addresses": [
            "192.168.122.42:24008"
        ],
        "client-addresses": [
            "127.0.0.1:24007",
            "192.168.122.42:24007"
        ],
        "online": true,
        "pid": 4896,
        "metadata": {
            "_zone": "854e037e-da80-4ee0-82b8-9ecdb8fc13de"
        }
    },
    {
        "id": "a35e9c07-fc7c-45ac-a0a3-4e09869b30ca",
        "name": "gd20.example",
        "peer-addresses": [
            "192.168.122.253:24008"
        ],
        "client-addresses": [
            "127.0.0.1:24007",
            "192.168.122.253:24007"
        ],
        "online": true,
        "pid": 4982,
        "metadata": {
            "_zone": "a35e9c07-fc7c-45ac-a0a3-4e09869b30ca"
        }
    }
]
```

###### Response
&nbsp;
Status : 500 Internal Server Error
```javascript
         {
             "errors": [
                 {
                     "code": 1,
                     "message": "failed to send leave cluster request"
                 }
             ]
         }
```
__Deletetion of Peer Results In Failure, So if anyone send a Get peers request, then it should present
in the peer list__

##### Response
&nbsp;
```
[
    {
        "id": "854e037e-da80-4ee0-82b8-9ecdb8fc13de",
        "name": "gd22.example",
        "peer-addresses": [
            "192.168.122.42:24008"
        ],
        "client-addresses": [
            "127.0.0.1:24007",
            "192.168.122.42:24007"
        ],
        "online": true,
        "pid": 4896,
        "metadata": {
            "_zone": "854e037e-da80-4ee0-82b8-9ecdb8fc13de"
        }
    },
    {
        "id": "a35e9c07-fc7c-45ac-a0a3-4e09869b30ca",
        "name": "gd20.example",
        "peer-addresses": [
            "192.168.122.253:24008"
        ],
        "client-addresses": [
            "127.0.0.1:24007",
            "192.168.122.253:24007"
        ],
        "online": true,
        "pid": 4982,
        "metadata": {
            "_zone": "a35e9c07-fc7c-45ac-a0a3-4e09869b30ca"
        }
    }
]
```

**Peer is not present in the list**

##### Logs
```
DEBU[2018-08-01 06:46:24.334238] received delete peer request                  peerid=2f0bfdac-fceb-459a-8939-689da17de77e reqid=d769751f-599e-4672-b873-cdc3a0ed752e source="[deletepeer.go:36:peers.deletePeerHandler]"
2018-08-01 06:46:24.336345 I | etcdserver/api/v3rpc: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 192.168.122.6:24008: connect: connection refused"; Reconnecting to {192.168.122.6:24008 <nil>}
ERRO[2018-08-01 06:46:24.336395] failed RPC call                               error="rpc error: code = 14 desc = grpc: the connection is unavailable" remote="192.168.122.6:24008" rpc=PeerService.Leave source="[peer-rpc-clnt.go:62:peers.(*peerSvcClnt).LeaveCluster]"
ERRO[2018-08-01 06:46:24.336445] sending Leave request failed                  error="rpc error: code = 14 desc = grpc: the connection is unavailable" peerid=2f0bfdac-fceb-459a-8939-689da17de77e reqid=d769751f-599e-4672-b873-cdc3a0ed752e source="[deletepeer.go:90:peers.deletePeerHandler]"
INFO[2018-08-01 06:46:24.336505] 192.168.122.1 - - [01/Aug/2018:06:46:24 -0400] "DELETE /v1/peers/2f0bfdac-fceb-459a-8939-689da17de77e HTTP/1.1" 500 73  reqid=d769751f-599e-4672-b873-cdc3a0ed752e
2018-08-01 06:46:24.336580 I | etcdserver/api/v3rpc: Failed to dial 192.168.122.6:24008: context canceled; please retry.

```

### PR DESCRIPTION
**Pull request by oshankkumar**
If Peer is not alive then sending a delete request for that peer will delete
Peer details from the store But Peer will fail to leave the cluster, Because
rpc call fails.

Before Deleting We should check whether peer is alive or not.
Fixes #1108

Signed-off-by: Oshank Kumar <okumar@redhat.com>

### CODE REVIEW

File: glusterd2/commands/peers/deletepeer.go [hunk 1]
```diff
@@ -43,13 +43,21 @@ func deletePeerHandler(w http.ResponseWriter, r *http.Request) {
...
+
 	// You cannot remove yourself
 	if id == gdctx.MyUUID.String() {
 		logger.Debug("request denied, received request to delete self from cluster")
 		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "removing self is disallowed.")
 		return
 	}
 
+	//Check if Node is alive
+	if _, alive := store.Store.IsNodeAlive(id); !alive {
```
Comment (aravindavk) [inline glusterd2/commands/peers/deletepeer.go]: alive is set by peer in regular interval. What happens to this request if alive state is set and then peer goes down and peer remove is called before the TTL of alive key?

Comment (oshankkumar) [reply glusterd2/commands/peers/deletepeer.go]: @aravindavk  then we can remove peer details from the store if a peer successfully left the cluster.

**Comment by oshankkumar** [comment]
retest this please

File: glusterd2/commands/peers/deletepeer.go [hunk 2]
```diff
@@ -61,13 +62,6 @@ func deletePeerHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
```
Comment (prashanthpai) [inline glusterd2/commands/peers/deletepeer.go]: Please reintroduce the `isAlive` soft check, in addition to moving `DeletePeer` below. That double check helps

Comment (oshankkumar) [reply glusterd2/commands/peers/deletepeer.go]: sure @prashanthpai

File: glusterd2/commands/peers/deletepeer.go [hunk 3]
```diff
@@ -43,13 +43,21 @@ func deletePeerHandler(w http.ResponseWriter, r *http.Request) {
...
 	// You cannot remove yourself
 	if id == gdctx.MyUUID.String() {
 		logger.Debug("request denied, received request to delete self from cluster")
 		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "removing self is disallowed.")
 		return
 	}
 
+	//Check if peer liveness key is present in store
+	if _, alive := store.Store.IsNodeAlive(id); !alive {
+		logger.Error("can not delete peer, peer is not alive")
```
Comment (Madhu-1) [inline glusterd2/commands/peers/deletepeer.go]: can you add peer id while logging? it will help for debugging

Comment (oshankkumar) [reply glusterd2/commands/peers/deletepeer.go]: It's already there in logger fields

**Pull request merged**